### PR TITLE
Build hnsw index

### PR DESCRIPTION
**Pull request by yun-zhou-rb**

### CODE REVIEW

File: pyrec/implicit/als.py [hunk 1]
```diff
@@ -85,7 +85,14 @@ def build_index(self, index_type: str, approximate_similar_items=True, approxima
                 **kwargs
             )
             return recommender
-
+        elif index_type == 'hnsw':
+            from .hnsw import ImplicitHNSWRecommender
```
Comment (vpillac) [inline pyrec/implicit/als.py:89]: why the `.` in `.hnsw`?

File: pyrec/implicit/als.py [hunk 2]
```diff
@@ -142,6 +148,15 @@ def load_recommender(als_model_file: str, index_file: str, item_feature_file: st
...
-
+    elif index_file.endswith('.hnsw'):
+        import hnswlib
+        from .hnsw import ImplicitHNSWRecommender
+        log.info("Loading hnsw recommendation index")
+        # we build the index in l2 space and load it in inner product space on purpose.
+        # This space change gives us 0.96 recall
+        l2_recommend_index = hnswlib.Index(space='ip', dim=model.item_factors.shape[1])
+        l2_recommend_index.load_index(index_file)
+        l2_recommend_index.set_ef(2000)
```
Comment (vpillac) [inline pyrec/implicit/als.py]: should that `2000`  be parametrisable?
you could add `**kwargs` to the method

Comment (vpillac) [reply pyrec/implicit/als.py]: Or should the caller then adjust the value?

**Review by vpillac — APPROVED** [review]

File: pyrec/implicit/als.py (hunk 1 above)
Comment (yun-zhou-rb) [reply pyrec/implicit/als.py:89]: relative path, from current folder, hnsw.py

File: pyrec/implicit/als.py (hunk 2 above)
Comment (yun-zhou-rb) [reply pyrec/implicit/als.py]: yes,  2000 can be parametrisable, I am just thinking set the ef once should be enough, not sure how many over header will set ef spend for every query.

**Review by vpillac — APPROVED** [review]

**Pull request merged**